controllers/awscluster_controller.go

```diff
@@ -0,0 +1,186 @@
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the AWSCluster object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
+func (r *AWSClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	log := log.FromContext(ctx)
+	log.Info("Started reconciling AWSCluster", "namespace", req.Namespace, "name", req.Name)
+	defer log.Info("Finished reconciling AWSCluster", "namespace", req.Namespace, "name", req.Name)
+
+	//
+	// Get AWSCluster that we are reconciling
+	//
+	awsCluster := &capa.AWSCluster{}
+	err := r.Client.Get(ctx, req.NamespacedName, awsCluster)
+	if err != nil {
+		return ctrl.Result{}, microerror.Mask(err)
+	}
+
+	// Check VPC mode. aws-vpc-operator reconciles only private VPCs.
+	vpcMode, vpcModeSet := awsCluster.Annotations[annotation.AWSVPCMode]
+	if !vpcModeSet || vpcMode != annotation.AWSVPCModePrivate {
+		return
+	}
+
+	// We don't reconcile AWSClusters that have VPC managed by CAPA
+	if awsCluster.Spec.NetworkSpec.VPC.IsManaged(awsCluster.Name) {
+		return ctrl.Result{}, nil
+	}
+
+	//
+	// Create patch helper that will update reconciler AWSCLuster if there are any changes in the CR
+	//
+	patchHelper, err := patch.NewHelper(awsCluster, r.Client)
+	if err != nil {
+		return ctrl.Result{}, microerror.Mask(err)
+	}
+	defer func() {
+		conditionsToUpdate := []capi.ConditionType{
+			capa.VpcReadyCondition,
+			// capa.SubnetsReadyCondition,
+			// capa.RouteTablesReadyCondition,
+		}
+		err := patchHelper.Patch(
+			ctx,
+			awsCluster,
+			patch.WithOwnedConditions{
+				Conditions: conditionsToUpdate,
+			})
+		if err != nil {
+			reterr = err
+		}
+	}()
+
+	// We need Spec.IdentityRef to be set, TODO check this
+	if awsCluster.Spec.IdentityRef == nil {
+		return ctrl.Result{}, microerror.Maskf(errors.IdentityNotSetError, "AWSCluster %s/%s does not have Spec.IdentityRef set", awsCluster.Namespace, awsCluster.Name)
+	}
+
+	identity := &capa.AWSClusterRoleIdentity{}
+	identityNamespacedName := types.NamespacedName{
+		Namespace: awsCluster.Namespace,
+		Name:      awsCluster.Spec.IdentityRef.Name,
+	}
+
+	err = r.Client.Get(ctx, identityNamespacedName, identity)
+	if err != nil {
+		return ctrl.Result{}, microerror.Mask(err)
+	}
+
+	vpcSpec := vpc.Spec{
+		ClusterName: awsCluster.Name,
+		RoleARN:     identity.Spec.RoleArn,
+		VpcId:       awsCluster.Spec.NetworkSpec.VPC.ID,
+		CidrBlock:   awsCluster.Spec.NetworkSpec.VPC.CidrBlock,
+	}
+	status, err := r.vpcReconciler.Reconcile(ctx, vpcSpec)
+	if err != nil {
+		return ctrl.Result{}, microerror.Mask(err)
+	}
```

Comment:
We should move the deferred function, there's still codepaths where errors can be returned rather than it succeeding.

```suggestion

	// We need Spec.IdentityRef to be set, TODO check this
	if awsCluster.Spec.IdentityRef == nil {
		return ctrl.Result{}, microerror.Maskf(errors.IdentityNotSetError, "AWSCluster %s/%s does not have Spec.IdentityRef set", awsCluster.Namespace, awsCluster.Name)
	}

	identity := &capa.AWSClusterRoleIdentity{}
	identityNamespacedName := types.NamespacedName{
		Namespace: awsCluster.Namespace,
		Name:      awsCluster.Spec.IdentityRef.Name,
	}

	err = r.Client.Get(ctx, identityNamespacedName, identity)
	if err != nil {
		return ctrl.Result{}, microerror.Mask(err)
	}

	defer func() {
		conditionsToUpdate := []capi.ConditionType{
			capa.VpcReadyCondition,
			// capa.SubnetsReadyCondition,
			// capa.RouteTablesReadyCondition,
		}
		err := patchHelper.Patch(
			ctx,
			awsCluster,
			patch.WithOwnedConditions{
				Conditions: conditionsToUpdate,
			})
		if err != nil {
			reterr = err
		}
	}()
	
	vpcSpec := vpc.Spec{
		ClusterName: awsCluster.Name,
		RoleARN:     identity.Spec.RoleArn,
		VpcId:       awsCluster.Spec.NetworkSpec.VPC.ID,
		CidrBlock:   awsCluster.Spec.NetworkSpec.VPC.CidrBlock,
	}
	status, err := r.vpcReconciler.Reconcile(ctx, vpcSpec)
	if err != nil {
		return ctrl.Result{}, microerror.Mask(err)
	}
```

Reply:
Fixed this in a different branch on top of this one.